### PR TITLE
MODEXPW-509 - Support Eureka permissions model for bulk operations (read operations)

### PR DESCRIPTION
**Pull request by alekGbuz**
[MODEXPW-509](https://folio-org.atlassian.net/browse/MODEXPW-509) - Support Eureka permissions model for bulk operations (read operations)
## Purpose

Implement logic to support permissions from mod-roles-keycloak and integrate it in existing approach.
## Approach

Introduce an environment variable in the module configuration that will be specifically depend to the environment to switch logic between mod-permissions and mod-roles-keycloak.
## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] Check logging
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Review by siarhei-charniak — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -59,3 +59,4 @@ This value should be `true` if AWS S3 is used as storage.
 | AUTHORITY_CONTROL_BATCH_JOB_CHUNK_SIZE            | 100                           | Specify chunk size for authority control export job which will be used to query data from entities-links, and write to file                                                                           |
 | AUTHORITY_CONTROL_BATCH_ENTITIES_LINKS_CHUNK_SIZE | 100                           | Amount to retrieve per request to mod-entities-links                                                                                                                                                  |
 | MAX_UPLOADED_FILE_SIZE                            | 40MB                          | Specifies multipart upload file size                                                                                                                                                                  |
+| EUREKA_PERMISSIONS_MODEL                          | false                         | Specifies if eureka permissions model is used                                                                                                                                                         |
```
Comment (khandramai) [inline README.md:62]: It is more informative to use env variable PLATFORM=okapi|eureka with default value okapi.

File: descriptors/ModuleDescriptor-template.json [hunk 2]
```diff
@@ -141,6 +141,10 @@
     {
       "id": "permissions",
       "version": "5.6"
+    },
+    {
+      "id": "permissions-users",
```
Comment (khandramai) [inline descriptors/ModuleDescriptor-template.json:146]: I assume permissions and permissions-users interfaces cannot be presented at the same time on platform. Probably, optional interface should be used here https://github.com/folio-org/okapi/blob/master/doc/guide.md#optional-interfaces.

File: descriptors/ModuleDescriptor-template.json [hunk 3]
```diff
@@ -276,7 +280,9 @@
           "methods": [ "GET" ],
           "pathPattern": "/bulk-edit/permissions-self-check",
           "permissionsRequired": ["bulk-edit.permissions-self-check.get"],
-          "modulePermissions": ["perms.users.get"]
+          "modulePermissions": [
+            "perms.users.get",
+            "permissions.users.item.get"]
```
Comment (khandramai) [inline descriptors/ModuleDescriptor-template.json:285]: Please, move brakes to the next line to follow common format of module descriptor.

**Review by obozhko-folio — APPROVED** [review]

File: src/main/java/org/folio/dew/service/UserPermissionsService.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+import java.util.List;
+
+
+@RequiredArgsConstructor
+@Log4j2
+@Service
+public class UserPermissionsService {
+
+  public static final String BULK_EDIT_INVENTORY_VIEW_PERMISSION = "bulk-operations.item.inventory.get";
```
Comment (khandramai) [inline src/main/java/org/folio/dew/service/UserPermissionsService.java:20]: Saving permissions in the enum looks more natural.

File: src/main/resources/application.yml [hunk 5]
```diff
@@ -112,6 +112,8 @@ application:
   chunks: ${CHUNKS:100}
   core-pool-size: ${CORE_POOL_SIZE:10}
   max-pool-size: ${MAX_POOL_SIZE:10}
+  eureka:
```
Comment (khandramai) [inline src/main/resources/application.yml:115]: platform: ${PLATFORM: okapi} looks more natural, see https://github.com/folio-org/mod-data-export-worker/pull/579/files#r1802609343

File: src/main/java/org/folio/dew/service/UserPermissionsService.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  private boolean isEurekaPermissionsModel;
+
+  private final FolioExecutionContext folioExecutionContext;
+  private final OkapiUserPermissionsClient okapiUserPermissionsClient;
+  private final EurekaUserPermissionsClient eurekaUserPermissionsClient;
+
+  public List<String> getPermissions() {
+    if (isEurekaPermissionsModel) {
+      var desiredPermissions = getDesiredPermissions();
+      return eurekaUserPermissionsClient.getPermissions(folioExecutionContext.getUserId().toString(),
```
Comment (khandramai) [inline src/main/java/org/folio/dew/service/UserPermissionsService.java:39]: Factory for getting permissions client can be used here: FeignClient permissionsClient = getPermissionsClient(String platform).

File: descriptors/ModuleDescriptor-template.json (hunk 2 above)
Comment (NikitaSedyx) [reply descriptors/ModuleDescriptor-template.json:146]: permissions is always present (Eureka uses it as well), but `permissions-users` is definitely should be moved to optional as okapi won't have it

File: src/main/resources/application.yml (hunk 5 above)
Comment (alekGbuz) [reply src/main/resources/application.yml]: Fixed

File: src/main/java/org/folio/dew/service/UserPermissionsService.java (hunk 4 above)
Comment (alekGbuz) [reply src/main/java/org/folio/dew/service/UserPermissionsService.java]: Updated

File: descriptors/ModuleDescriptor-template.json (hunk 3 above)
Comment (alekGbuz) [reply descriptors/ModuleDescriptor-template.json]: Updated

File: descriptors/ModuleDescriptor-template.json (hunk 2 above)
Comment (alekGbuz) [reply descriptors/ModuleDescriptor-template.json:148]: It was done thanks

File: README.md (hunk 1 above)
Comment (alekGbuz) [reply README.md]: Updated

File: src/main/java/org/folio/dew/service/UserPermissionsService.java (hunk 6 above)
Comment (alekGbuz) [reply src/main/java/org/folio/dew/service/UserPermissionsService.java:42]: Hi @khandramai , the arguments of clients and return value is different, from my point of view it is possible to use as is.

**Comment by alekGbuz** [comment]
Hi @khandramai please take a look.

**Review by khandramai — APPROVED** [review]

**Comment by khandramai** [comment]
> Hi @khandramai please take a look on updates.

LGTM, go ahead. Thanks.

**Pull request merged**